Resources/ProxyProvider/DIRACCAProxyProvider.py

```diff
@@ -33,227 +66,249 @@ def __init__(self, parameters=None):
     self.match = []
     self.supplied = ['CN']
     self.optional = ['C', 'O', 'OU', 'emailAddress']
-    # Add not supported distributes names
-    self.fs2nid = X509.X509_Name.nid.copy()
-    self.fs2nid['DC'] = -1
-    self.fs2nid['domainComponent'] = -1
-    self.fs2nid['organizationalUnitName'] = 18
-    self.fs2nid['countryName'] = 14
-    self.n2field = {}  # nid: most short or specidied in CS distributes name
-    self.n2fields = {}  # nid: list of distributes names
+    self.dnList = ['C', 'O', 'OU', 'CN', 'emailAddress']
+    # Distinguished names
+    self.fields2nid = X509.X509_Name.nid.copy()
+    self.fields2nid['DC'] = -1  # Add DN that is not liested in X509.X509_Name
+    self.fields2nid['domainComponent'] = -1  # Add DN description that is not liested in X509.X509_Name
+    self.fields2nid['organizationalUnitName'] = 18  # Add 'OU' description
+    self.fields2nid['countryName'] = 14  # Add 'C' description
+    self.fields2nid['SERIALNUMBER'] = 105  # Add 'SERIALNUMBER' distinguished name
+    self.nid2field = {}  # nid: most short or specidied in CS distinguished name
+    self.nid2fields = {}  # nid: list of distinguished names
     # Specify standart fields
-    for field in self.fs2nid:
-      if self.fs2nid[field] not in self.n2fields:
-        self.n2fields[self.fs2nid[field]] = []
-      self.n2fields[self.fs2nid[field]].append(field)
-    for nid in self.n2fields:
-      for field in self.n2fields[nid]:
-        if nid not in self.n2field:
-          self.n2field[nid] = field
-        self.n2field[nid] = len(field) < len(self.n2field[nid]) and field or self.n2field[nid]
+    for field in self.fields2nid:
+      if self.fields2nid[field] not in self.nid2fields:
+        self.nid2fields[self.fields2nid[field]] = []
+      self.nid2fields[self.fields2nid[field]].append(field)
+    for nid in self.nid2fields:
+      for field in self.nid2fields[nid]:
+        if nid not in self.nid2field:
+          self.nid2field[nid] = field
+        self.nid2field[nid] = len(field) < len(self.nid2field[nid]) and field or self.nid2field[nid]
+    self.dnInfoDictCA = {}
 
   def setParameters(self, parameters):
     """ Set new parameters
 
         :param dict parameters: provider parameters
+
+        :return: S_OK()/S_ERROR()
     """
     # If CA configuration file exist
     self.parameters = parameters
     if parameters.get('CAConfigFile'):
-      self.supplied, self.optional = [], []
       self.__parseCACFG()
     if 'Bits' in parameters:
       self.bits = int(parameters['Bits'])
     if 'Algoritm' in parameters:
       self.algoritm = parameters['Algoritm']
     if 'Match' in parameters:
       self.match = []
-      for field in parameters['Match'].replace(' ', '').split(','):
-        self.match.append(self.fs2nid[field])
+      if not isinstance(parameters['Match'], list):
+        parameters['Match'] = parameters['Match'].replace(', ', ',').split(',')
```

Comment:
you are doing the same with `Match`, `Supplied`, `Optional`, `DNOrder`, etc. You could factorize all these loops

Reply:
done